File: packages/host-node/src/http/assets.ts
/**
 * Static assets: the packaged web build, and nothing else.
 *
 * The HTTP layer serves exactly one directory — the web bundle the CLI was shipped
 * with. It never serves a repository, a Git directory, a log or a configuration
 * file, and it never builds a path from client input without proving the result is
 * inside that directory.
 *
 * The rules, and the failure each one prevents:
 *
 * - **No traversal.** The decoded, normalised path must stay inside the asset root,
 *   compared after `realpath` so a symlink inside the bundle cannot point at `/etc`.
 * - **No NUL and no encoded separators.** `%00`, `%2f` and a literal `\0` are
 *   refused rather than passed to the filesystem, where a NUL truncates a path and
 *   an encoded slash can defeat a naive prefix check.
 * - **A missing asset is a 404, not the SPA.** The app shell may answer for a route
 *   (`/repos/x`), but `/assets/app-123.js` must never return HTML: a browser that
 *   receives HTML where it expected JavaScript reports a syntax error instead of a
 *   missing file, and a service worker would cache the wrong thing.
 * - **`/api/*` never falls through.** An unknown API path is a JSON 404, so a
 *   client cannot mistake an HTML shell for an API answer.
 * - **CSP and friends.** The page loads only its own bytes: no remote script, no
 *   CDN, no remote font, no framing by another origin. A document's own inline
 *   bootstrap script is named by hash rather than permitted in general (see `csp.ts`),
 *   which is what lets a static single-page app boot without `'unsafe-inline'`.
 *   An embedding static host may add exact API origins to `connect-src`; the API's
 *   own origin and bearer checks remain separate and authoritative.
 */
import { createReadStream } from "node:fs";
import { readFile, realpath, stat } from "node:fs/promises";
import { extname, join, normalize, sep } from "node:path";
import type { ServerResponse } from "node:http";
import type { Problem } from "@refyard/git-contract";
import { inlineScriptHashes, policyWithInlineScripts } from "./csp.js";

export interface AssetServer {
  /** Serve one path, or report why it cannot be. */
  serve(input: {
    readonly path: string;
    readonly response: ServerResponse;
    readonly headOnly: boolean;
  }): Promise<"served" | "not-found" | "refused">;
  root(): string | null;
}

export interface AssetServerOptions {
  /** Directory of the built web app; when absent, no static asset can be served. */
  readonly webRoot: string | null;
  /** Fallback document for client-side routes, e.g. `200.html`. */
  readonly fallbackDocument?: string;
  /**
   * A document served when there is no web build.
   *
   * The CLI ships before the Svelte app does, and a service that answers every
   * browser request with 503 looks broken. This page says what is true — the API is
   * running, the UI is not in this build — and is served only for document-shaped
   * paths, never for a missing asset or an API path.
  */
  readonly inlineDocument?: string;
  /** Exact API origins an embedding static host needs to call from its own origin. */
  readonly connectOrigins?: readonly string[];
}

const CONTENT_TYPES: Readonly<Record<string, string>> = {
  ".html": "text/html; charset=utf-8",
  ".js": "text/javascript; charset=utf-8",
  ".mjs": "text/javascript; charset=utf-8",
  ".css": "text/css; charset=utf-8",
  ".json": "application/json; charset=utf-8",
  ".map": "application/json; charset=utf-8",
  ".svg": "image/svg+xml",
  ".png": "image/png",
  ".jpg": "image/jpeg",
  ".jpeg": "image/jpeg",
  ".webp": "image/webp",
  ".ico": "image/x-icon",
  ".woff": "font/woff",
  ".woff2": "font/woff2",
  ".txt": "text/plain; charset=utf-8",
  ".webmanifest": "application/manifest+json",
};

/**
 * Headers every HTML response carries.
 *
 * `default-src 'self'` plus `script-src 'self'` is what makes "no external CDN
 * script, no remote font" enforceable rather than aspirational; `object-src 'none'`
 * and `frame-ancestors 'none'` close the two ways a page can be made to load
 * something else. The default `connect-src 'self'` keeps the API same-origin, which
 * is also why there is no CORS header anywhere in this service. An embedding host can
 * opt into exact API origins without changing that default.
 *
 * Author avatars are the one cross-origin image: GitHub serves the photo for a
 * noreply commit email, fetched with no referrer and no credentials.
 */
export const HTML_HEADERS: Readonly<Record<string, string>> = {
  "content-type": "text/html; charset=utf-8",
  "cache-control": "no-store",
  "x-content-type-options": "nosniff",
  "referrer-policy": "no-referrer",
  "content-security-policy":
    "default-src 'self'; script-src 'self'; style-src 'self' 'unsafe-inline'; img-src 'self' data: https://github.com https://avatars.githubusercontent.com; font-src 'self'; connect-src 'self'; object-src 'none'; base-uri 'none'; frame-ancestors 'none'; form-action 'none'",
  "x-frame-options": "DENY",
};

export function createAssetServer(options: AssetServerOptions): AssetServer {
  const fallback = options.fallbackDocument ?? "200.html";
  const htmlHeaders = headersWithConnectOrigins(options.connectOrigins);
  let resolvedRoot: string | null = null;
  let rootReady = false;
  /**
   * Per-document headers, keyed by the file's identity and not by its path.
   *
   * The policy names the hash of the document's own inline scripts, so it is only correct
   * for the bytes it was computed from. A packaged bundle does not change while the
   * service runs — but a development build does: `pnpm build` rewrites these files under
   * a running service, and a cache keyed by path alone would keep answering with the old
   * hash. The browser would then refuse the new document's inline bootstrap, and the app
   * would never start. mtime and size are what the filesystem changes on a rewrite.
   */
  const documentHeaderCache = new Map<
    string,
    {
      readonly mtimeMs: number;
      readonly size: number;
      readonly headers: Readonly<Record<string, string>>;
    }
  >();

  async function documentHeaders(
    absolutePath: string,
  ): Promise<Readonly<Record<string, string>>> {
    let stamp: { readonly mtimeMs: number; readonly size: number } | null =
      null;
    try {
      const info = await stat(absolutePath);
      stamp = { mtimeMs: info.mtimeMs, size: info.size };
    } catch {
      // A document that cannot be stat-ed cannot be cached; the read below reports the
      // same failure with the strict policy.
      stamp = null;
    }
    const cached = documentHeaderCache.get(absolutePath);
    if (
      cached !== undefined &&
      stamp !== null &&
      cached.mtimeMs === stamp.mtimeMs &&
      cached.size === stamp.size
    ) {
      return cached.headers;
    }
    let headers: Readonly<Record<string, string>> = htmlHeaders;
    try {
      const html = await readFile(absolutePath, "utf8");
      const hashes = inlineScriptHashes(html);
      headers =
        hashes.length === 0
          ? HTML_HEADERS
          : {
              ...htmlHeaders,
              "content-security-policy": policyWithInlineScripts(
                htmlHeaders["content-security-policy"] ?? "",
                hashes,
              ),
            };
    } catch {
      // A document that cannot be read is served (or fails) with the strict policy; a
      // policy is not the place to report an I/O error.
      headers = htmlHeaders;
    }
    if (stamp !== null) {
      documentHeaderCache.set(absolutePath, { ...stamp, headers });
    }
    return headers;
  }

  /**
   * The resolved asset root, memoised until a request misses because of it.
   *
   * `refresh` drops the memo and resolves again. It is used once per request that could
   * not be answered — see `serve` — because `realpath` is what turns a symlinked web root
   * into an absolute path, and a rebuild that retargets the link (or replaces the
   * directory) leaves this process holding the old one forever: the app would answer
   * "not found" for its own shell until someone restarted the service.
   */
  async function assetRoot(refresh = false): Promise<string | null> {
    if (rootReady && !refresh) {
      return resolvedRoot;
    }
    rootReady = true;
    if (options.webRoot === null) {
      resolvedRoot = null;
      return null;
    }
    const previous = resolvedRoot;
    try {
      const info = await stat(options.webRoot);
      resolvedRoot = info.isDirectory()
        ? await realpath(options.webRoot)
        : null;
    } catch {
      resolvedRoot = null;
    }
    if (refresh && resolvedRoot === previous) {
      // Nothing changed: the miss was a genuine 404, not a stale root.
      return resolvedRoot;
    }
    return resolvedRoot;
  }

  async function resolveAsset(
    path: string,
  ): Promise<
    | { readonly kind: "file"; readonly absolutePath: string }
    | { readonly kind: "not-found" }
    | { readonly kind: "refused"; readonly problem: Problem }
  > {
    const root = await assetRoot();
    if (root === null) {
      return { kind: "not-found" };
    }
    if (path.includes("\u0000") || path.includes("%00")) {
      return {
        kind: "refused",
        problem: {
          code: "InvalidRequest",
          message: "the path contains a NUL byte",
          retryable: false,
        },
      };
    }
    let decoded: string;
    try {
      decoded = decodeURIComponent(path);
    } catch {
      return {
        kind: "refused",
        problem: {
          code: "InvalidRequest",
          message: "the path is not valid percent-encoding",
          retryable: false,
        },
      };
    }
    if (
      decoded.includes("\u0000") ||
      decoded.includes(sep === "/" ? "\u0000" : "\\\u0000")
    ) {
      return {
        kind: "refused",
        problem: {
          code: "InvalidRequest",
          message: "the decoded path contains a NUL byte",
          retryable: false,
        },
      };
    }
    const candidate = normalize(join(root, decoded));
    // Lexical containment first (cheap, and catches `..`), then a realpath check
    // when the file exists (catches a symlink out of the bundle).
    if (candidate !== root && !candidate.startsWith(root + sep)) {
      return {
        kind: "refused",
        problem: {
          code: "Forbidden",
          message: "that path is outside the web assets",
          retryable: false,
        },
      };
    }
    try {
      const info = await stat(candidate);
      if (!info.isFile()) {
        return { kind: "not-found" };
      }
      const real = await realpath(candidate);
      if (real !== root && !real.startsWith(root + sep)) {
        return {
          kind: "refused",
          problem: {
            code: "Forbidden",
            message: "that asset resolves outside the web assets",
            retryable: false,
          },
        };
      }
      return { kind: "file", absolutePath: real };
    } catch {
      return { kind: "not-found" };
    }
  }

  async function sendFile(input: {
    readonly absolutePath: string;
    readonly response: ServerResponse;
    readonly headOnly: boolean;
    readonly isDocument: boolean;
  }): Promise<void> {
    const extension = extname(input.absolutePath).toLowerCase();
    const headers = input.isDocument
      ? await documentHeaders(input.absolutePath)
      : {
          "content-type":
            CONTENT_TYPES[extension] ?? "application/octet-stream",
          // Hashed asset names change when their contents change, so a short cache
          // is safe; the document itself is always revalidated.
          "cache-control": "public, max-age=300",
          "x-content-type-options": "nosniff",
        };
    for (const [name, value] of Object.entries(headers)) {
      input.response.setHeader(name, value);
    }
    if (input.headOnly) {
      input.response.writeHead(200);
      input.response.end();
      return;
    }
    input.response.writeHead(200);
    await new Promise<void>((resolve, reject) => {
      const stream = createReadStream(input.absolutePath);
      stream.on("error", reject);
      stream.on("end", resolve);
      stream.pipe(input.response);
    });
  }

  return {
    root(): string | null {
      return resolvedRoot;
    },

    async serve({
      path,
      response,
      headOnly,
    }): Promise<"served" | "not-found" | "refused"> {
      let resolved = await resolveAsset(path);
      if (resolved.kind === "not-found") {
        // The root is resolved once per process, so a rebuild that replaced it — or a
        // symlink that was retargeted at the new build — would leave this service
        // answering "not found" for its own shell until someone restarted it. One retry
        // after re-resolving is the difference between "the UI is gone" and "the next
        // request lands in the new build", and it costs a `stat` on a miss.
        const before = resolvedRoot;
        const after = await assetRoot(true);
        if (after !== null && after !== before) {
          resolved = await resolveAsset(path);
        }
      }
      if (resolved.kind === "refused") {
        response.setHeader("content-type", "application/json; charset=utf-8");
        response.setHeader("cache-control", "no-store");
        response.writeHead(
          resolved.problem.code === "Forbidden"
            ? 403
            : resolved.problem.code === "InvalidRequest"
              ? 400
              : 503,
        );
        response.end(JSON.stringify({ problem: resolved.problem }));
        return "refused";
      }
      if (resolved.kind === "file") {
        await sendFile({
          absolutePath: resolved.absolutePath,
          response,
          headOnly,
          isDocument: extname(resolved.absolutePath).toLowerCase() === ".html",
        });
        return "served";
      }

      // Not a file. A path that looks like an asset (or an API call) is a 404; only
      // a route-shaped path falls back to the app shell.
      const looksLikeAsset = /\.[A-Za-z0-9]+$/.test(path);
      if (looksLikeAsset || path.startsWith("/api/")) {
        return "not-found";
      }
      const root = await assetRoot();
      if (root === null) {
        if (options.inlineDocument === undefined) {
          return "not-found";
        }
        // No web build in this installation: answer a document request with a page
        // that says exactly that, instead of a 503 that reads like a failure.
        for (const [name, value] of Object.entries(htmlHeaders)) {
          response.setHeader(name, value);
        }
        response.writeHead(200);
        if (headOnly) {
          response.end();
        } else {
          response.end(options.inlineDocument);
        }
        return "served";
      }
      const fallbackPath = join(root, fallback);
      try {
        const info = await stat(fallbackPath);
        if (!info.isFile()) {
          return "not-found";
        }
      } catch {
        return "not-found";
      }
      await sendFile({
        absolutePath: fallbackPath,
        response,
        headOnly,
        isDocument: true,
      });
      return "served";
    },
  };
}

/** Add only exact configured API origins to the static document's connect policy. */
function headersWithConnectOrigins(
  origins: readonly string[] | undefined,
): Readonly<Record<string, string>> {
  const valid = [
    ...new Set((origins ?? []).filter((origin) => isHttpOrigin(origin))),
  ];
  if (valid.length === 0) {
    return HTML_HEADERS;
  }
  const policy = HTML_HEADERS["content-security-policy"] ?? "";
  return {
    ...HTML_HEADERS,
    "content-security-policy": policy.replace(
      "connect-src 'self'",
      `connect-src 'self' ${valid.join(" ")}`,
    ),
  };
}

function isHttpOrigin(value: string): boolean {
  try {
    const url = new URL(value);
    return (
      (url.protocol === "http:" || url.protocol === "https:") &&
      url.origin === value
    );
  } catch {
    return false;
  }
}

File: crates/refyard-http/src/assets.rs
//! The static workbench: one directory of built assets, served literally.
//!
//! The rules are stricter than a general web server's because the audience is a browser
//! that will run what it is handed:
//!
//! - **one root, no escape.** A request path is decoded, joined under the root, and both
//!   the lexical join and the resolved path are checked to stay inside it — a symlinked
//!   asset that points out of the tree is refused, not followed.
//! - **the SPA fallback is shape-dependent.** A path that looks like an asset (it has a
//!   file extension) or starts with `/api/` is answered 404 when missing — never with
//!   HTML, which would turn a broken asset link into a broken page that loads. Only a
//!   route-shaped path falls back to `200.html`, the pre-rendered shell.
//! - **documents carry a restrictive CSP**, and because the built shell has exactly one
//!   inline `<script>` (SvelteKit's boot data), each document's inline script bodies are
//!   hashed into `script-src` at serve time — an attacker cannot inject a second inline
//!   script even though one is allowed to run.

use std::path::{Path, PathBuf};

use refyard_contract::problem::{Problem, ProblemCode};

/// The pre-rendered shell a route-shaped request falls back to.
const FALLBACK_DOCUMENT: &str = "200.html";

/// What a request for a static path produced.
#[derive(Debug)]
pub enum Serving {
    /// Bytes to send with these headers.
    File {
        body: Vec<u8>,
        headers: Vec<(&'static str, String)>,
    },
    /// The path looks like an asset or an API call, and nothing is there: an API/asset
    /// 404, never the shell.
    NotFound,
}

/// Serves files from `root`, or reports what the caller should answer instead.
pub fn serve(root: &Path, request_path: &str, head_only: bool) -> Result<Serving, Problem> {
    if request_path.contains('\0') || request_path.contains("%00") {
        return Err(Problem::new(
            ProblemCode::InvalidRequest,
            "the path contains a NUL byte",
        ));
    }
    let decoded = percent_decode(request_path)?;
    if decoded.contains('\0') {
        return Err(Problem::new(
            ProblemCode::InvalidRequest,
            "the path contains a NUL byte",
        ));
    }

    let root = root
        .canonicalize()
        .map_err(|_| Problem::new(ProblemCode::Unavailable, "the web assets are missing"))?;
    let relative = decoded.trim_start_matches('/');
    let candidate = normalize_join(&root, relative);
    if !candidate.starts_with(&root) {
        return Err(Problem::new(
            ProblemCode::Forbidden,
            "that path is outside the web assets",
        ));
    }
    let resolved = candidate.canonicalize().ok();
    if resolved
        .as_ref()
        .is_some_and(|path| !path.starts_with(&root))
    {
        return Err(Problem::new(
            ProblemCode::Forbidden,
            "that asset resolves outside the web assets",
        ));
    }

    let path = resolved.unwrap_or(candidate);
    if path.is_file() {
        let body = std::fs::read(&path)
            .map_err(|_| Problem::new(ProblemCode::Unavailable, "the asset could not be read"))?;
        return Ok(Serving::File {
            headers: headers_for(&path, &body),
            body: if head_only { Vec::new() } else { body },
        });
    }

    // Missing. Asset-shaped and API-shaped paths stay 404; only a route gets the shell.
    let asset_shaped = Path::new(&decoded).extension().is_some_and(|extension| {
        !extension.is_empty() && is_asset_extension(&extension.to_string_lossy())
    });
    if asset_shaped || decoded.starts_with("/api") {
        return Ok(Serving::NotFound);
    }
    let shell = root.join(FALLBACK_DOCUMENT);
    let body = std::fs::read(&shell)
        .map_err(|_| Problem::new(ProblemCode::Unavailable, "the web shell is missing"))?;
    Ok(Serving::File {
        headers: headers_for(&shell, &body),
        body: if head_only { Vec::new() } else { body },
    })
}

/// Joins a decoded request path under the root, resolving `.` and `..` lexically.
///
/// `Path::join` keeps `..` as a literal component and `starts_with` compares components
/// literally, so a `..` path would look in-root right up until the filesystem resolved
/// it. Normalising here makes the lexical check honest before the filesystem is asked
/// anything.
fn normalize_join(root: &Path, relative: &str) -> PathBuf {
    use std::ffi::OsString;
    let mut components: Vec<OsString> = root
        .components()
        .map(|component| component.as_os_str().to_owned())
        .collect();
    for part in relative.split('/') {
        match part {
            "" | "." => {}
            ".." => {
                components.pop();
            }
            other => components.push(OsString::from(other)),
        }
    }
    components.into_iter().collect()
}

/// The response headers for one file: caching for assets, a strict document policy for
/// HTML.
fn headers_for(path: &Path, body: &[u8]) -> Vec<(&'static str, String)> {
    let is_html = path
        .extension()
        .is_some_and(|extension| extension == "html");
    if is_html {
        vec![
            ("content-type", "text/html; charset=utf-8".to_string()),
            ("cache-control", "no-store".to_string()),
            ("x-content-type-options", "nosniff".to_string()),
            ("referrer-policy", "no-referrer".to_string()),
            ("content-security-policy", content_security_policy(body)),
            ("x-frame-options", "DENY".to_string()),
        ]
    } else {
        vec![
            ("content-type", content_type(path).to_string()),
            ("cache-control", "public, max-age=300".to_string()),
            ("x-content-type-options", "nosniff".to_string()),
        ]
    }
}

/// The document policy. Every directive narrows what a page may load or do; `connect-src`
/// is `'self'` because the workbench talks to the origin that served it and nowhere else.
fn content_security_policy(document: &[u8]) -> String {
    let mut policy = String::from("default-src 'self'; script-src 'self'");
    for hash in inline_script_hashes(document) {
        policy.push_str(&format!(" 'sha256-{hash}'"));
    }
    policy.push_str(
        "; style-src 'self' 'unsafe-inline'; img-src 'self' data:; font-src 'self'; \
         connect-src 'self'; object-src 'none'; base-uri 'none'; frame-ancestors 'none'; \
         form-action 'none'",
    );
    policy
}

/// The standard-base64 SHA-256 of every inline `<script>` body in a document.
///
/// A scanner rather than an HTML parser, on purpose: the input is this repository's own
/// build output, not the web at large, and the scanner states what it accepts — script
/// tags whose opening tag contains no `src=` — which is exactly the shape SvelteKit's
/// prerenderer emits.
fn inline_script_hashes(document: &[u8]) -> Vec<String> {
    use sha2::{Digest, Sha256};
    let text = String::from_utf8_lossy(document);
    let lowercase = text.to_lowercase();
    let mut hashes = Vec::new();
    let mut cursor = 0usize;
    while let Some(open) = lowercase[cursor..].find("<script") {
        let open = cursor + open;
        let Some(tag_end) = lowercase[open..].find('>') else {
            break;
        };
        let tag = &text[open..open + tag_end];
        let body_starts = open + tag_end + 1;
        if tag.to_lowercase().contains("src=") {
            // An external script loads under `script-src 'self'`; nothing to hash.
        } else if let Some(close) = lowercase[body_starts..].find("</script>") {
            let body = &text[body_starts..body_starts + close];
            let digest = Sha256::digest(body.as_bytes());
            hashes.push(base64_standard(&digest));
            cursor = body_starts + close;
            continue;
        } else {
            break;
        }
        cursor = body_starts;
    }
    hashes
}

fn is_asset_extension(extension: &str) -> bool {
    extension
        .chars()
        .all(|character| character.is_ascii_alphanumeric())
}

/// The media type of an asset, by extension, exactly as the build output names them.
fn content_type(path: &Path) -> &'static str {
    let extension = path
        .extension()
        .map(|extension| extension.to_string_lossy().to_lowercase())
        .unwrap_or_default();
    match extension.as_str() {
        "html" => "text/html; charset=utf-8",
        "js" | "mjs" => "text/javascript; charset=utf-8",
        "css" => "text/css; charset=utf-8",
        "json" | "map" => "application/json; charset=utf-8",
        "svg" => "image/svg+xml",
        "png" => "image/png",
        "jpg" | "jpeg" => "image/jpeg",
        "webp" => "image/webp",
        "ico" => "image/x-icon",
        "woff" => "font/woff",
        "woff2" => "font/woff2",
        "txt" => "text/plain; charset=utf-8",
        "webmanifest" => "application/manifest+json",
        _ => "application/octet-stream",
    }
}

/// Strict percent-decoding: an escape this decoder cannot read is an error, not garbage
/// in a filename.
fn percent_decode(input: &str) -> Result<String, Problem> {
    let bytes = input.as_bytes();
    let mut output = Vec::with_capacity(bytes.len());
    let mut index = 0;
    while index < bytes.len() {
        if bytes[index] == b'%' {
            let hex = bytes.get(index + 1..index + 3).ok_or_else(|| {
                Problem::new(
                    ProblemCode::InvalidRequest,
                    "the path is not valid percent-encoding",
                )
            })?;
            let decoded = u8::from_str_radix(
                std::str::from_utf8(hex).map_err(|_| {
                    Problem::new(
                        ProblemCode::InvalidRequest,
                        "the path is not valid percent-encoding",
                    )
                })?,
                16,
            )
            .map_err(|_| {
                Problem::new(
                    ProblemCode::InvalidRequest,
                    "the path is not valid percent-encoding",
                )
            })?;
            output.push(decoded);
            index += 3;
        } else {
            output.push(bytes[index]);
            index += 1;
        }
    }
    String::from_utf8(output)
        .map_err(|_| Problem::new(ProblemCode::InvalidRequest, "the path is not valid UTF-8"))
}

/// Standard base64 with padding — the alphabet a CSP hash is written in.
fn base64_standard(input: &[u8]) -> String {
    const ALPHABET: &[u8; 64] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
    let mut output = String::with_capacity(input.len().div_ceil(3) * 4);
    for chunk in input.chunks(3) {
        let bytes = [
            chunk[0],
            *chunk.get(1).unwrap_or(&0),
            *chunk.get(2).unwrap_or(&0),
        ];
        let number = ((bytes[0] as u32) << 16) | ((bytes[1] as u32) << 8) | bytes[2] as u32;
        output.push(ALPHABET[(number >> 18) as usize & 63] as char);
        output.push(ALPHABET[(number >> 12) as usize & 63] as char);
        output.push(if chunk.len() > 1 {
            ALPHABET[(number >> 6) as usize & 63] as char
        } else {
            '='
        });
        output.push(if chunk.len() > 2 {
            ALPHABET[number as usize & 63] as char
        } else {
            '='
        });
    }
    output
}

#[cfg(test)]
mod tests {
    use super::*;

    fn root() -> tempfile::TempDir {
        let temp = tempfile::tempdir().expect("temp dir");
        let web = temp.path().join("web");
        std::fs::create_dir_all(&web).expect("mkdir web");
        std::fs::write(
            web.join("200.html"),
            "<!doctype html><html><body><script>let boot = {\"a\":1};</script><script src=\"/x.js\"></script></body></html>",
        )
        .expect("write shell");
        std::fs::create_dir_all(web.join("_app")).expect("mkdir");
        std::fs::write(web.join("_app").join("app.js"), "export const x = 1;\n").expect("write js");
        // A file next to the web root, so an escape attempt has something to reach for
        // without writing outside this test's own directory. The directory is returned
        // held: dropping it would delete the tree the assertions have not read yet.
        std::fs::write(temp.path().join("secret.txt"), b"nope").expect("write secret");
        temp
    }

    /// The served directory inside a fixture that is kept alive by the test.
    fn web(temp: &std::path::Path) -> PathBuf {
        temp.join("web")
    }

    #[test]
    fn a_route_shaped_request_falls_back_to_the_shell_and_hashes_its_inline_script() {
        let temp = root();
        let serving = serve(&web(temp.path()), "/repositories/some/id", false).expect("serves");
        let Serving::File { headers, .. } = serving else {
            panic!("the shell is served");
        };
        let csp = headers
            .iter()
            .find(|(name, _)| *name == "content-security-policy")
            .map(|(_, value)| value.clone())
            .expect("a CSP");
        assert!(csp.contains("script-src 'self"), "{csp}");
        // SHA-256 of `let boot = {"a":1};`, standard base64.
        assert!(csp.contains("'sha256-"), "{csp}");
        assert!(csp.contains("frame-ancestors 'none'"), "{csp}");
    }

    #[test]
    fn a_missing_asset_stays_404_and_never_becomes_the_shell() {
        let temp = root();
        assert!(matches!(
            serve(&web(temp.path()), "/_app/missing.js", false),
            Ok(Serving::NotFound)
        ));
        assert!(matches!(
            serve(&web(temp.path()), "/api/v1/whatever", false),
            Ok(Serving::NotFound)
        ));
    }

    #[test]
    fn an_existing_asset_is_served_with_its_type_and_a_short_cache() {
        let temp = root();
        let serving = serve(&web(temp.path()), "/_app/app.js", false).expect("serves");
        let Serving::File { headers, body } = serving else {
            panic!("the asset is served");
        };
        assert_eq!(body, b"export const x = 1;\n");
        assert!(headers.contains(&("content-type", "text/javascript; charset=utf-8".to_string())));
        assert!(headers.contains(&("cache-control", "public, max-age=300".to_string())));
    }

    // Prevents: `/../../etc/passwd` or a symlink planted in the tree reading a file the
    // server never meant to serve.
    #[test]
    fn a_path_that_escapes_the_root_is_refused() {
        let temp = root();
        for attempt in [
            "/../secret.txt",
            "/%2e%2e/secret.txt",
            "/_app/../../secret.txt",
        ] {
            let problem = serve(&web(temp.path()), attempt, false).expect_err(attempt);
            assert_eq!(problem.code, ProblemCode::Forbidden, "{attempt}");
        }
    }

    #[test]
    fn a_nul_byte_or_broken_escape_is_an_invalid_path() {
        let temp = root();
        for attempt in ["/%00", "/a%zz", "/a%2"] {
            let problem = serve(&web(temp.path()), attempt, false).expect_err(attempt);
            assert_eq!(problem.code, ProblemCode::InvalidRequest, "{attempt}");
        }
    }

    #[test]
    fn a_head_request_returns_headers_with_no_body() {
        let temp = root();
        let serving = serve(&web(temp.path()), "/_app/app.js", true).expect("serves");
        let Serving::File { body, .. } = serving else {
            panic!("the asset is served");
        };
        assert!(body.is_empty());
    }

    #[test]
    fn standard_base64_is_the_alphabet_a_csp_hash_is_written_in() {
        assert_eq!(base64_standard(b"foobar"), "Zm9vYmFy");
        assert_eq!(base64_standard(b"foob"), "Zm9vYg==");
        use sha2::{Digest, Sha256};
        let empty = base64_standard(&Sha256::digest(b""));
        assert_eq!(empty, "47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU=");
    }
}
